File: internal/tg/announce.go
package tg

import (
	"context"
	"log/slog"
	"sort"
	"strings"
	"sync"
	"time"
	"unicode/utf8"

	"github.com/go-telegram/bot"
	"github.com/go-telegram/bot/models"
)

// An announcement is still just a message handed back to Telegram by id, so
// this stays agnostic about how its content is encoded: it needs enough text to
// judge the length, and the ids to send.
const (
	minAnnounceLength = 200
	// Telegram delivers an album as separate updates; they land within a few
	// milliseconds of each other, so a short wait collects the whole set.
	albumDebounce = time.Second

	announcePrompt = "Пришлите сообщение для отправки в канал с анонсами:"
	tooShort       = "Слишком короткое сообщение. Введите новую версию или нажмите /cancel"
	unparseable    = "Не получилось распарсить сообщение :( Введите новую версию или нажмите /cancel"
	announceSent   = "Анонс успешно отправлен!"
	announceFailed = "Что-то пошло не так :( Напишите разработчику бота"
)

type albumKey struct {
	chatID  int64
	groupID string
}

type albumBuffer struct {
	debounce time.Duration

	mu     sync.Mutex
	groups map[albumKey][]*models.Message
	timers map[albumKey]*time.Timer
	// Stragglers of an album already relayed must not start a second post.
	recent map[albumKey]bool
	order  []albumKey
}

func newAlbumBuffer() *albumBuffer {
	return &albumBuffer{
		debounce: albumDebounce,
		groups:   map[albumKey][]*models.Message{},
		timers:   map[albumKey]*time.Timer{},
		recent:   map[albumKey]bool{},
	}
}

func (a *albumBuffer) markSent(key albumKey) {
	a.recent[key] = true
	a.order = append(a.order, key)
	if len(a.order) > 64 {
		delete(a.recent, a.order[0])
		a.order = a.order[1:]
	}
}

func (b *Bot) announceEntry(ctx context.Context, message *models.Message) {
	// Set the state before the round-trip: updates are handled concurrently,
	// so a fast follow-up can arrive while the prompt is still in flight.
	if err := b.fsm.SetName(b.key(message), stateAnnounce); err != nil {
		slog.Error("saving conversation state", "err", err)
	}
	b.reply(ctx, message, announcePrompt)
}

func (b *Bot) announce(ctx context.Context, message *models.Message) {
	if message.MediaGroupID == "" {
		b.deliverAnnounce(ctx, message, []*models.Message{message})
		return
	}
	b.collectAlbum(message)
}

func (b *Bot) collectAlbum(message *models.Message) {
	key := albumKey{chatID: message.Chat.ID, groupID: message.MediaGroupID}
	a := b.albums
	a.mu.Lock()
	defer a.mu.Unlock()

	if a.recent[key] {
		slog.Info("ignoring album item arriving after the album was relayed",
			"chat_id", key.chatID, "album", key.groupID)
		return
	}
	a.groups[key] = append(a.groups[key], message)
	if timer, waiting := a.timers[key]; waiting {
		timer.Stop()
	}
	a.timers[key] = time.AfterFunc(a.debounce, func() { b.flushAlbum(key) })
}

func (b *Bot) flushAlbum(key albumKey) {
	a := b.albums
	a.mu.Lock()
	group := a.groups[key]
	delete(a.groups, key)
	delete(a.timers, key)
	if len(group) == 0 {
		a.mu.Unlock()
		return
	}
	a.markSent(key)
	a.mu.Unlock()

	sort.Slice(group, func(i, j int) bool { return group[i].ID < group[j].ID })
	// The organiser may have cancelled while the debounce was pending.
	if b.fsm.Get(b.key(group[0])).Name != stateAnnounce {
		slog.Info("dropping album, conversation already left", "chat_id", key.chatID)
		return
	}
	b.deliverAnnounce(context.Background(), group[0], group)
}

func (b *Bot) deliverAnnounce(ctx context.Context, message *models.Message, group []*models.Message) {
	var parts []string
	messageIDs := make([]int, 0, len(group))
	for _, item := range group {
		if part := MessageText(item); part != "" {
			parts = append(parts, part)
		}
		messageIDs = append(messageIDs, item.ID)
	}
	body := strings.Join(parts, "\n")

	if body == "" {
		b.reply(ctx, message, unparseable)
		return
	}
	if utf8.RuneCountInString(body) < minAnnounceLength {
		b.reply(ctx, message, tooShort)
		return
	}

	chatID := message.Chat.ID
	isForward := group[0].ForwardOrigin != nil
	sent := b.relay(ctx, chatID, messageIDs, isForward)
	if !sent && !isForward {
		slog.Info("copy refused, forwarding instead", "message_ids", messageIDs)
		sent = b.relay(ctx, chatID, messageIDs, true)
	}

	if sent {
		b.reply(ctx, message, announceSent)
		slog.Info("announce sent", "chat_id", chatID, "text", body)
	} else {
		b.reply(ctx, message, announceFailed)
		slog.Error("could not send announce", "chat_id", chatID)
	}
	if err := b.fsm.Clear(b.key(message)); err != nil {
		slog.Error("clearing conversation state", "err", err)
	}
}

// relay hands the message back to Telegram by id, copying it when it is the
// organiser's own and forwarding it when it is not.
func (b *Bot) relay(ctx context.Context, fromChatID int64, messageIDs []int, forward bool) bool {
	target := b.config.AnnounceChannelID
	method, err := b.callRelay(ctx, target, fromChatID, messageIDs, forward)
	b.logs.RecordTraffic("out", method, target, "", jsonOf(map[string]any{
		"chat_id": target, "from_chat_id": fromChatID, "message_ids": messageIDs,
	}))
	if err != nil {
		slog.Error("relaying announce", "method", method, "message_ids", messageIDs,
			"target", target, "err", err)
		return false
	}
	return true
}

func (b *Bot) callRelay(
	ctx context.Context, target, fromChatID int64, messageIDs []int, forward bool,
) (string, error) {
	if len(messageIDs) > 1 {
		if forward {
			_, err := b.api.ForwardMessages(ctx, &bot.ForwardMessagesParams{
				ChatID: target, FromChatID: fromChatID, MessageIDs: messageIDs})
			return "forwardMessages", err
		}
		_, err := b.api.CopyMessages(ctx, &bot.CopyMessagesParams{
			ChatID: target, FromChatID: fromChatID, MessageIDs: messageIDs})
		return "copyMessages", err
	}
	if forward {
		_, err := b.api.ForwardMessage(ctx, &bot.ForwardMessageParams{
			ChatID: target, FromChatID: fromChatID, MessageID: messageIDs[0]})
		return "forwardMessage", err
	}
	_, err := b.api.CopyMessage(ctx, &bot.CopyMessageParams{
		ChatID: target, FromChatID: fromChatID, MessageID: messageIDs[0]})
	return "copyMessage", err
}

File: src/rating_bot/handlers/announce.py
"""Relaying an organiser's announcement into the announce channel.

Whatever Telegram invents next, an announcement is still just a message we hand
back to Telegram by id, so this stays agnostic about how the content is encoded
-- it only needs enough text to judge the length, and the ids to send.
"""

import asyncio
import logging
from collections import deque

from aiogram import Bot, Router
from aiogram.filters import Command
from aiogram.fsm.context import FSMContext
from aiogram.types import Message

from rating_bot.config import Config
from rating_bot.handlers.common import reply_html
from rating_bot.helpers import udumps
from rating_bot.rich_text import message_text
from rating_bot.states import Flows

logger = logging.getLogger("rating_bot")

router = Router(name="announce")

MIN_LENGTH = 200
# Telegram delivers an album as separate updates; they land within a few ms of
# each other, so a short wait collects the whole set.
ALBUM_DEBOUNCE_SECONDS = 1.0

PROMPT = "Пришлите сообщение для отправки в канал с анонсами:"
TOO_SHORT = "Слишком короткое сообщение. Введите новую версию или нажмите /cancel"
UNPARSEABLE = "Не получилось распарсить сообщение :( Введите новую версию или нажмите /cancel"
SENT = "Анонс успешно отправлен!"
FAILED = "Что-то пошло не так :( Напишите разработчику бота"

_albums: dict[tuple[int, str], list[Message]] = {}
_album_timers: dict[tuple[int, str], asyncio.Task] = {}
# Stragglers of an album we already relayed must not start a second post.
_recently_sent: deque[tuple[int, str]] = deque(maxlen=64)


@router.message(Command("announce"))
async def announce_entry(message: Message, state: FSMContext) -> None:
    # Set the state before the round-trip: updates are handled concurrently, so
    # a fast follow-up message can arrive while the prompt is still in flight.
    await state.set_state(Flows.announce)
    await reply_html(message, PROMPT)


@router.message(Flows.announce)
async def announce(
    message: Message, state: FSMContext, bot: Bot, config: Config, **_
) -> None:
    if message.media_group_id:
        await _collect_album(message, state, bot, config)
        return
    await _deliver(message, [message], state, bot, config)


async def _collect_album(
    message: Message, state: FSMContext, bot: Bot, config: Config
) -> None:
    key = (message.chat.id, message.media_group_id)
    if key in _recently_sent:
        logger.info(f"ignoring album item arriving after {key} was relayed")
        return
    _albums.setdefault(key, []).append(message)
    if timer := _album_timers.get(key):
        timer.cancel()

    async def flush() -> None:
        # Cancellation means a later item of this album superseded us; the
        # buffer belongs to its timer now, so unwind without touching it.
        await asyncio.sleep(ALBUM_DEBOUNCE_SECONDS)
        try:
            group = sorted(_albums.get(key, []), key=lambda m: m.message_id)
            if not group:
                return
            # The organiser may have cancelled while the debounce was pending.
            if await state.get_state() != Flows.announce.state:
                logger.info(f"dropping album {key}, conversation already left")
                return
            _recently_sent.append(key)
            await _deliver(group[0], group, state, bot, config)
        except Exception as e:
            logger.error(f"exception {type(e)} {e} while sending album announce")
        finally:
            _albums.pop(key, None)
            _album_timers.pop(key, None)

    _album_timers[key] = asyncio.create_task(flush())


async def _deliver(
    message: Message,
    group: list[Message],
    state: FSMContext,
    bot: Bot,
    config: Config,
) -> None:
    text = "\n".join(filter(None, (message_text(item) for item in group)))
    if not text:
        await reply_html(message, UNPARSEABLE)
        return
    if len(text) < MIN_LENGTH:
        await reply_html(message, TOO_SHORT)
        return

    chat_id = message.chat.id
    message_ids = [item.message_id for item in group]
    is_forward = group[0].forward_origin is not None

    sent = await _send(bot, config, chat_id, message_ids, forward=is_forward)
    if not sent and not is_forward:
        logger.info(f"copy refused for {message_ids}, forwarding instead")
        sent = await _send(bot, config, chat_id, message_ids, forward=True)

    if sent:
        await reply_html(message, SENT)
        logger.info(f"user {chat_id} sent announce {udumps(text)}")
    else:
        await reply_html(message, FAILED)
        logger.error(f"error while trying to send announce from user {chat_id}")
    await state.clear()


async def _send(
    bot: Bot, config: Config, from_chat_id: int, message_ids: list[int], *, forward: bool
) -> bool:
    target = config.announce_channel_id
    try:
        if len(message_ids) > 1:
            send = bot.forward_messages if forward else bot.copy_messages
            await send(
                chat_id=target, from_chat_id=from_chat_id, message_ids=message_ids
            )
        else:
            send = bot.forward_message if forward else bot.copy_message
            await send(
                chat_id=target, from_chat_id=from_chat_id, message_id=message_ids[0]
            )
        return True
    except Exception as e:
        logger.error(
            f"exception {type(e)} {e} while trying to"
            f" {'forward' if forward else 'copy'} {message_ids} to {target}"
        )
        return False
